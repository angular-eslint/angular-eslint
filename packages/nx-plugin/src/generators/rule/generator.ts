import { formatFiles, generateFiles, logger } from '@nx/devkit';
import { type Tree } from '@nx/devkit';
import * as path from 'path';
import { type RuleGeneratorSchema } from './schema';

export async function ruleGenerator(tree: Tree, options: RuleGeneratorSchema) {
  const { ruleName, packageName } = options;
  const ruleNameCamelCase = ruleName.replace(/-./g, (c) => c[1].toUpperCase());
  const baseTarget = `packages/${packageName}`;

  generateFiles(
    tree,
    path.join(__dirname, 'files/src'),
    `${baseTarget}/src`,
    options,
  );
  generateFiles(
    tree,
    path.join(__dirname, 'files/tests'),
    `${baseTarget}/tests/rules/${ruleName}`,
    options,
  );

  // Update package's index.ts to export the new rule
  const packageIndexContents = tree
    .read(`${baseTarget}/src/index.ts`)
    ?.toString() as string;
  const importString = `import ${ruleNameCamelCase}, { RULE_NAME as ${ruleNameCamelCase}RuleName } from './rules/${ruleName}';`;
  let newPackageIndexContents = packageIndexContents.replace(
    '\nexport = {',
    `${importString}\n\nexport = {`,
  );
  newPackageIndexContents = newPackageIndexContents.replace(
    'rules: {',
    `rules: { [${ruleNameCamelCase}RuleName]: ${ruleNameCamelCase},`,
  );
  tree.write(`${baseTarget}/src/index.ts`, newPackageIndexContents);

  await formatFiles(tree);

  return () => {
    logger.log(
      `\nSkeleton for new rule generated successfully. See TODO comments for areas to complete.`,
    );
    logger.log(
      `Don't forget to run \`update-configs\` and \`update-rule-docs\` after implementing the rule and its tests!`,
    );
  };
}

export default ruleGenerator;
