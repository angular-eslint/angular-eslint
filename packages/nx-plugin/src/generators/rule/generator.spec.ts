import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { type Tree } from '@nx/devkit';

import { ruleGenerator } from './generator';
import type { RuleGeneratorSchema } from './schema';

describe('rule generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });

  it('should successfully generate a new rule for package eslint-plugin', async () => {
    const options: RuleGeneratorSchema = {
      ruleName: 'my-new-rule',
      packageName: 'eslint-plugin',
    };
    const packageIndexPath = 'packages/eslint-plugin/src/index.ts';
    const mockIndex = `
      \nexport = {
        configs: {},
        rules: {},
      };
    `;

    tree.write(packageIndexPath, mockIndex);
    await ruleGenerator(tree, options);

    expect(
      tree.exists(`packages/eslint-plugin/src/rules/my-new-rule.ts`),
    ).toBeTruthy();
    expect(
      tree.exists(`packages/eslint-plugin/tests/rules/my-new-rule/cases.ts`),
    ).toBeTruthy();
    expect(
      tree.exists(`packages/eslint-plugin/tests/rules/my-new-rule/spec.ts`),
    ).toBeTruthy();

    const packageIndexContents = tree
      .read(packageIndexPath)
      ?.toString() as string;

    expect(
      packageIndexContents.includes(
        `import myNewRule, { RULE_NAME as myNewRuleRuleName } from './rules/my-new-rule';`,
      ),
    ).toBeTruthy();
    expect(
      packageIndexContents.includes(`[myNewRuleRuleName]: myNewRule`),
    ).toBeTruthy();
  });

  it('should successfully generate a new rule for package eslint-plugin-template', async () => {
    const options: RuleGeneratorSchema = {
      ruleName: 'my-new-rule',
      packageName: 'eslint-plugin-template',
    };
    const packageIndexPath = 'packages/eslint-plugin-template/src/index.ts';
    const mockIndex = `
      \nexport = {
        configs: {},
        rules: {},
      };
    `;

    tree.write(packageIndexPath, mockIndex);
    await ruleGenerator(tree, options);

    expect(
      tree.exists(`packages/eslint-plugin-template/src/rules/my-new-rule.ts`),
    ).toBeTruthy();
    expect(
      tree.exists(
        `packages/eslint-plugin-template/tests/rules/my-new-rule/cases.ts`,
      ),
    ).toBeTruthy();
    expect(
      tree.exists(
        `packages/eslint-plugin-template/tests/rules/my-new-rule/spec.ts`,
      ),
    ).toBeTruthy();

    const packageIndexContents = tree
      .read(packageIndexPath)
      ?.toString() as string;

    expect(
      packageIndexContents.includes(
        `import myNewRule, { RULE_NAME as myNewRuleRuleName } from './rules/my-new-rule';`,
      ),
    ).toBeTruthy();
    expect(
      packageIndexContents.includes(`[myNewRuleRuleName]: myNewRule`),
    ).toBeTruthy();
  });
});
