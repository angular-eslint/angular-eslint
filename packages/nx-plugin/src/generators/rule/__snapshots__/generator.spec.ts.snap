// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule generator should successfully generate a new rule for package eslint-plugin 1`] = `
"import { createESLintRule } from '../utils/create-eslint-rule';

export type Options = [];
export type MessageIds = 'myNewRule';
export const RULE_NAME = 'my-new-rule';

export default createESLintRule<Options, MessageIds>({
  name: RULE_NAME,
  // TODO: Complete metadata, fixing any type errors
  meta: {
    type: '',
    docs: {
      description: '',
      recommended: '',
    },
    // fixable: '', TODO: Is it fixable?
    schema: [],
    messages: {
      myNewRule: '',
    }
  },
  defaultOptions: [],
  create(context) {
    // TODO: return RuleListener
  },
});
"
`;

exports[`rule generator should successfully generate a new rule for package eslint-plugin 2`] = `
"import { convertAnnotatedSourceToFailureCase } from '@angular-eslint/test-utils';
import type { InvalidTestCase, ValidTestCase } from '@typescript-eslint/rule-tester';
import type { MessageIds, Options } from '../../../src/rules/my-new-rule';

const messageId: MessageIds = 'myNewRule';

export const valid: readonly (string | ValidTestCase<Options>)[] = [
  // TODO: Add valid cases
];

export const invalid: readonly InvalidTestCase<MessageIds, Options>[] = [
  // TODO: Add invalid cases
];
"
`;

exports[`rule generator should successfully generate a new rule for package eslint-plugin 3`] = `
"import { RuleTester } from '@angular-eslint/test-utils';
import rule, { RULE_NAME } from '../../../src/rules/my-new-rule';
import { invalid, valid } from './cases';

const ruleTester = new RuleTester();

ruleTester.run(RULE_NAME, rule, {
  valid,
  invalid,
});
"
`;
